File: android/app/src/main/java/com/example/tokn/TokenWidgetProvider.java
package com.example.tokn;
import com.example.tokn.R;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;




import es.antonborri.home_widget.HomeWidgetPlugin;

/**
 * TokN Home Screen Widget – shows upcoming booking token details.
 */
public class TokenWidgetProvider extends AppWidgetProvider {

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        String hospitalName  = HomeWidgetPlugin.Companion.getData(context).getString("hospital", "City General");
        String servingNumber = HomeWidgetPlugin.Companion.getData(context).getString("serving",  "42");
        String mineNumber    = HomeWidgetPlugin.Companion.getData(context).getString("mine",     "48");
        String waitTime      = HomeWidgetPlugin.Companion.getData(context).getString("wait_time", "15");

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_layout);
        views.setTextViewText(R.id.widget_hospital,    hospitalName);
        views.setTextViewText(R.id.widget_serving_num, "#" + servingNumber);
        views.setTextViewText(R.id.widget_mine_num,    "#" + mineNumber);
        views.setTextViewText(R.id.widget_wait_time,   "~" + waitTime + " mins wait");

        // Open App on click
        Intent intent = new Intent(context, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        views.setOnClickPendingIntent(R.id.widget_root, pendingIntent);

        appWidgetManager.updateAppWidget(appWidgetId, views);

    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }
}

File: android/app/src/main/kotlin/com/example/tokn/TokenWidgetProvider.kt
package com.example.tokn

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.SharedPreferences
import android.widget.RemoteViews
import es.antonborri.home_widget.HomeWidgetProvider

class TokenWidgetProvider : HomeWidgetProvider() {
    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray, widgetData: SharedPreferences) {
        for (appWidgetId in appWidgetIds) {
            val views = RemoteViews(context.packageName, R.layout.widget_layout).apply {
                // Get data from Flutter HomeWidget
                val hospitalName = widgetData.getString("hospital", "No Hospital") ?: "No Hospital"
                val servingNum = widgetData.getString("serving", "-") ?: "-"
                val mineNum = widgetData.getString("mine", "-") ?: "-"
                val waitTime = widgetData.getString("wait_time", "0") ?: "0"

                // Update UI views
                setTextViewText(R.id.widget_hospital, hospitalName)
                setTextViewText(R.id.widget_serving_num, "#$servingNum")
                setTextViewText(R.id.widget_mine_num, "#$mineNum")
                setTextViewText(R.id.widget_wait_time, "~$waitTime mins wait")
            }

            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }
}
